services/paymaster/src/db/config.rs

```diff
@@ -0,0 +1,177 @@
+          v.instructions  AS instructions
+        FROM domain_config d
+        JOIN variation v ON v.domain_config_id = d.id
+        ORDER BY d.id
+        "#,
+    )
+    .fetch_all(pool::pool())
+    .await?;
+
+    let mut map: HashMap<Uuid, Domain> = HashMap::new();
+
+    for r in rows {
+        map.entry(r.domain_id)
+            .or_insert_with(|| Domain {
+                domain: r.domain,
+                enable_session_management: r.enable_session_management,
+                enable_preflight_simulation: r.enable_preflight_simulation,
+                tx_variations: Vec::new(),
+            })
+            .tx_variations
+            .push(r.instructions.0);
+    }
+
+    Ok(Config {
+        domains: map.into_values().collect(),
+    })
+}
+
+/// Hash a plaintext password using Argon2.
+fn hash_password(plain: &str) -> String {
+    let salt = SaltString::generate(&mut OsRng);
+
+    // Argon2 with default params (Argon2id v19)
+    let argon2 = Argon2::default();
+
+    // Hash password to PHC string ($argon2id$v=19$...)
+    argon2
+        .hash_password(plain.as_bytes(), &salt)
+        .unwrap()
+        .to_string()
+}
+
+// Get the registrable domain from a URL.
+// Example: https://test.brasa.finance -> brasa.finance
+fn registrable_domain(u: &Url) -> Option<String> {
+    let host = u.host_str()?;
+    let parts: Vec<&str> = host.split('.').collect();
+    if parts.len() >= 2 {
+        Some(format!(
+            "{}.{}",
+            parts[parts.len() - 2],
+            parts[parts.len() - 1]
+        ))
+    } else {
+        Some(host.to_string())
+    }
+}
+/// Insert a user(email, password) into the database.
+async fn insert_user(domain_url: &Url) -> Result<Uuid, sqlx::Error> {
+    let password = hash_password("admin");
+    let email = format!("admin@{}", &registrable_domain(domain_url).unwrap());
+    let existing_user = sqlx::query_as::<_, (Uuid,)>("SELECT id FROM \"user\" WHERE email = $1")
+        .bind(&email)
+        .fetch_optional(pool())
+        .await?;
+
+    if let Some(user_result) = existing_user {
+        return Ok(user_result.0);
+    }
+
+    let new_user = sqlx::query_as::<_, (Uuid,)>(
+        "INSERT INTO \"user\" (email, password) VALUES ($1, $2) RETURNING id",
+    )
+    .bind(&email)
+    .bind(&password)
+    .fetch_one(pool())
+    .await?;
+
+    Ok(new_user.0)
+}
+
+async fn insert_app(user_id: &Uuid, name: &str) -> Result<Uuid, sqlx::Error> {
+    let app = sqlx::query_as::<_, (Uuid,)>(
+        "INSERT INTO app (user_id, name) VALUES ($1, $2) RETURNING id",
+    )
+    .bind(user_id)
+    .bind(name)
+    .fetch_one(pool())
+    .await?;
+
+    Ok(app.0)
+}
+
+async fn insert_domain_config(app_id: &Uuid, domain: &Domain) -> Result<Uuid, sqlx::Error> {
+    let domain_config = sqlx::query_as::<_, (Uuid,)>(
+        "INSERT INTO domain_config (app_id, domain, enable_session_management, enable_preflight_simulation) VALUES ($1, $2, $3, $4) RETURNING id",
```

Comment:
if you run sqlx::prepare, it should generate a cached query statement which optimizes running this query with different params.